core/lib/tests/drop-in-async-context.rs

```diff
@@ -0,0 +1,63 @@
+#[macro_use]
+extern crate rocket;
+
+use rocket::{build, fairing::AdHoc, Build, Orbit, Rocket};
+
+struct AsyncDropInAsync;
+
+impl Drop for AsyncDropInAsync {
+    fn drop(&mut self) {
+        // Attempt to fetch the current runtime while dropping
+        // Pools in rocket_sync_db_pools (and maybe rocket_db_pools)
+        // do use this capability. They spawn tasks to asyncronously
+        // complete shutdown of the pool, which triggers the same panic.
+        let _ = rocket::tokio::runtime::Handle::current();
+    }
+}
+
+fn rocket() -> Rocket<Build> {
+    build().manage(AsyncDropInAsync).attach(AdHoc::on_liftoff(
+        "Shutdown immediately",
+        |rocket: &Rocket<Orbit>| {
+            Box::pin(async {
+                rocket.shutdown().notify();
+            })
+        },
+    ))
+}
+
+mod launch {
+    #[launch]
+    fn launch() -> _ {
+        super::rocket()
+    }
+    #[test]
+    fn test_launch() {
+        main();
+    }
+}
+
+mod main {
+    use rocket::tokio::net::TcpListener;
+
+    #[rocket::main]
+    async fn main() {
+        super::rocket()
+            .try_launch_on(TcpListener::bind("localhost:8001"))
+            .await
+            .unwrap();
+    }
+    #[test]
+    fn test_main() {
+        main();
+    }
+    #[test]
+    fn test_execute() {
+        rocket::execute(async {
+            super::rocket()
+                .try_launch_on(TcpListener::bind("localhost:8002"))
+                .await
+                .unwrap();
+        });
+    }
+}
```

Comment:
This is extremely fragile. The only correct way to start a TCP server in tests is to do so with an unnumbered port.

Given what this code is doing, I _really_ believe this needs to be a test in `testbench`. The tests are easier to write this way, too. All you need to do is introduce a new module in [testbench/src/servers](https://github.com/rwf2/Rocket/tree/db598be3a8e7ed3d5a01d8ef5eddfd6868c22333/testbench/src/servers). The returned handle allows you to shutdown the server as well.

Reply:
Yeah, I should switch to using an unnumbered port.

That being said, I'm not confident I can use testbench for this specific case - at least not without some issues. The challenge is that I'm trying to test the `#[launch]` macro itself - it looks like most cases pass a rocket instance to the spawn macro, which then attaches a fairing. I can't do that - since I'm directly executing the generated (sync) main function.;

Reply:
I've taken the time to make an attempt (it required some modifications to testbench, and still doesn't work). You can check it out from the 'drop-in-async-launch-testbench' branch of my fork.